streamingccl: add replication lag to `SHOW VIRTUAL CLUSTER` results

msbutler opened the pull request:
With this patch, SHOW VIRTUAL CLUSTER ... WITH REPLICATION STATUS will display the replication lag.

Fixes #120647

Release note (sql change): SHOW VIRTUAL CLUSTER ... WITH REPLICATION STATUS now displays the PCR replication lag.

msbutler committed:
streamingccl: add replication lag to `SHOW VIRTUAL CLUSTER` results

With this patch, SHOW VIRTUAL CLUSTER ... WITH REPLICATION STATUS will display
the replication lag.

Fixes #120647

Release note (sql change): SHOW VIRTUAL CLUSTER ... WITH REPLICATION STATUS now
displays the PCR replication lag.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120782)
<!-- Reviewable:end -->

dt commented on pkg/sql/catalog/colinfo/result_columns.go:
```diff
@@ -305,6 +305,7 @@ var TenantColumnsWithReplication = ResultColumns{
 	{Name: "source_tenant_name", Typ: types.String},
 	{Name: "source_cluster_uri", Typ: types.String},
 	{Name: "replication_job_id", Typ: types.Int},
```
I chatted with @alicia-l2 yesterday and we were we would remove replication_job_id to make space for the new lag column. She also thought it would be more intuitive to order it as retained_time, replicated_time, replication_lag, cutover_time

msbutler replied:
Done.

stevendanna replied:
Hrm. Are we sure about this?  For instance if you find your replication job PAUSED or FAILED, how confident are we that in this command we always show the relevant error output? I suppose `SELECT * FROM [SHOW JOBS] WHERE job_type = 'STREAM INGESTION'` isn't too hard. but it is a bit more than copy/pasting the job ID and doing `SHOW JOB`.

dt replied:
> how confident are we that in this command we always show the relevant error output

This sounds like something _we_ should be on the hook to find/fix. Needing to go find the job ID in `show jobs` or the db console is _slightly_ annoying sure, but if it is just something that we would be doing when debugging a bug and then we go fix the status update code, that seems like the right trade for the end-user UX being that we show them just the things that are relevant _to them_, which is the replication status?

dt replied:
One observation is that this table is _much_ easier to read when it fits on one line. Once it wraps it is much harder to line up which timestamp your'e looking at with the headers. The job_id's cost, in line width and noise, to most users seems higher than the benefit give our not-job-centric UX here, so if it is just there for debugging by CRL engineers, I think we can say we're on the hook for finding it the slightly harder way to reclaim the line width.

stevendanna replied:
OK. I'll follow up next week with a test that makes sure we get the job error in the status output. I feel OK arguing any fix we need there is worth backporting.

dt replied:
(... in the `with replication status` version) and rename `data_state` to `status` and remove `service_mode` from this version while we're at it?

stevendanna replied:
> Should we move the status column all the way to the right while we're here in case there is a long error in it?

I think we should.  We'll need to reach out to preview users who have queries against this already, but that is fine.

> (... in the with replication status version) and rename data_state to status and remove service_mode from this version while we're at it?

I think in the original command we were thinking that _WITH_ would only _add_ columns.  But, if we are making breaking changes we shouldn't be shy.

msbutler replied:
just pushed up another commit. if CI is pleased, I'll squash and merge.

msbutler committed:
streamingccl: refactor SHOW VIRTUAL CLUSTER ... with REPLICATION STATUS

Epic: none

Release note (sql change): This patch refactors the output of SHOW VC WITH
REPLICATION STATUS to 1) remove the replication_job_id, service_mode return
fields; 2) reorder to following columns to: retained_time, replicated_time,
replication_lag, cutover_time; 3) rename the data_state field status and moves
to the last return column.

dt commented on pkg/sql/show_tenant.go:
```diff
@@ -242,7 +244,12 @@ func (n *showTenantNode) Values() tree.Datums {
+				minIngestedMicro := minIngested.GoTime().Truncate(time.Microsecond)
+				replicatedTimestamp, _ = tree.MakeDTimestampTZ(minIngestedMicro, time.Nanosecond)
+
+				nowMicro := timeutil.Now().Truncate(time.Microsecond)
```
is p.EvalContext().GetStmtTimestamp more correct than observing current wall time?

msbutler replied:
good point. done. note that I had to add another field to `showTenantNode`

dt commented on pkg/sql/show_tenant.go:
```diff
@@ -70,13 +73,13 @@ func (p *planner) ShowTenant(ctx context.Context, n *tree.ShowTenant) (planNode,
 		withPriorReplication: n.WithPriorReplication,
 		withCapabilities:     n.WithCapabilities,
 		initTenantValues:     true,
+		statementTime:        p.EvalContext().GetStmtTimestamp(),
```
hm, I think this is being evaluated at prep, rather than exec, now? is there not an eval ctx you can use during exec?

dt replied:
I don't know for sure, but I think the runParams to startExec are what you're supposed to use to evaluate expressions during exec

msbutler replied:
moved the call to exec. unclear if it makes a difference.

dt commented:
this looks like it needs a `./dev testlogic base --files tenant --rewrite`

msbutler commented:
TFTRs!

bors r=dt
